consensus/src/dag.rs

```diff
@@ -1,28 +1,352 @@
+    ),
+}
+
+impl<PublicKey: VerifyingKey> InnerDag<PublicKey> {
+    pub fn spawn(rx_primary: Receiver<Certificate<PublicKey>>) -> (JoinHandle<()>, Dag<PublicKey>) {
+        let (tx_commands, rx_commands) = tokio::sync::mpsc::channel(DEFAULT_CHANNEL_SIZE);
+
+        let handle = tokio::spawn(async move {
+            Self {
+                rx_primary,
+                rx_commands,
+                dag: NodeDag::new(),
+                vertices: BTreeMap::new(),
+            }
+            .run()
+            .await
+        });
+        let dag = Dag { tx_commands };
+        (handle, dag)
     }
 
     async fn run(&mut self) {
-        // at the moment just receive the certificate and throw away
-        while self.rx_primary.recv().await.is_some() {
-            debug!("Received certificate, will ignore");
+        loop {
+            tokio::select! {
+                 Some(certificate) = self.rx_primary.recv() => {
+                    // The Core (process_certificate) guarantees the certificate
+                    // has gone through causal completion => this is ready to be inserted
+                    let _ = self.insert(certificate);
+                }
+                Some(command) = self.rx_commands.recv() => {
+                    match command {
+                        DagCommand::Insert(cert, sender) => { let _ = sender.send(self.insert(cert)); },
+                        DagCommand::Contains(dig, sender)=> {
+                            let _ = sender.send(self.contains(dig));
+                        },
+                        DagCommand::Rounds(pk, sender) => {
+                            let _ = sender.send(self.rounds(pk));
+                        },
+                        DagCommand::Remove(dig, sender) => {
+                            let _ = sender.send(self.remove(dig));
+                        },
+                        DagCommand::ReadCausal(dig, sender) => {
+                            let res = self.read_causal(dig);
+                            let _ = sender.send(res.map(|r| r.collect()));
+                        },
+                        DagCommand::NodeReadCausal((pk, round), sender) => {
+                            let res = self.node_read_causal(pk, round);
+                            let _ = sender.send(res.map(|r| r.collect()));
+                        },
+                    }
+                }
+            }
+        }
+    }
+
+    #[instrument(level = "debug", err)]
+    fn insert(
+        &mut self,
+        certificate: Certificate<PublicKey>,
+    ) -> Result<(), ValidatorDagError<PublicKey>> {
+        let digest = certificate.digest();
+        let round = certificate.round();
+        let origin = certificate.origin();
+
+        // This fails if the validation of the certificate is incomplete
+        self.dag.try_insert(certificate)?;
+        // TODO: atomicity with the previous
+        self.vertices.insert((origin, round), digest);
```

Comment:
Since we extract `(origin, round)` from the certificate itself, and its hash calculation covers over those elements, short of finding a collision on blake2, we will insert the exact same record. I don't see grounds for rejection.
On  the Dag, the re-insertion will be (precisely) a no-op (it won't revive a compressed vertex, for instance).